lib/certificate_chain_serializer.rb

```diff
@@ -0,0 +1,18 @@
+class CertificateChainSerializer
+  PATTERN = /-----BEGIN CERTIFICATE-----(?:.|\n)+?-----END CERTIFICATE-----/.freeze
+
+  def self.load(chain)
+    return [] unless chain
+    chain.scan(PATTERN).map do |cert|
+      OpenSSL::X509::Certificate.new(cert)
+    end
+  end
+
+  def self.dump(chain)
+    return if chain.blank?
+    normalised = chain.map do |cert|
+      cert.respond_to?(:to_pem) ? cert : OpenSSL::X509::Certificate.new(cert)
```

Comment:
can you please explain a scenario where cert may not respond to `to_pem`?

Reply:
That's there because `Gem::Specification#cert_chain` returns an array of strings instead of an array of `OpenSSL::X509::Certificate` objects…

Reply:
okay, that is the expected case then. when will we `cert` as is (without Cert.new)? I see that it is used in tests. perhaps we should make tests match the expected case where cert is an array of strings.
Also, Cert.new is blowing up for invalid cert.
```
Version.create(number: "0.0.1", platform: "ruby", cert_chain: ["notcert"])
  TRANSACTION (0.2ms)  BEGIN
  Version Exists? (0.3ms)  SELECT 1 AS one FROM "versions" WHERE "versions"."full_name" IS NULL LIMIT $1  [["LIMIT", 1]]
  TRANSACTION (0.2ms)  ROLLBACK
Traceback (most recent call last):
        7: from (irb):3
        6: from (irb):4:in `rescue in irb_binding'
        5: from lib/certificate_chain_serializer.rb:13:in `dump'
        4: from lib/certificate_chain_serializer.rb:13:in `map'
        3: from lib/certificate_chain_serializer.rb:14:in `block in dump'
        2: from lib/certificate_chain_serializer.rb:14:in `new'
        1: from lib/certificate_chain_serializer.rb:14:in `initialize'
OpenSSL::X509::CertificateError (not enough data)

```

Reply:
```
2.7.4 :007 > Version.create(number: "0.0.1", platform: "ruby", cert_chain: "/search?utf8=%E2%9C%93&query=php://input%00.&search_submit=%E2%8C%95" )
  TRANSACTION (0.3ms)  BEGIN
  Version Exists? (0.5ms)  SELECT 1 AS one FROM "versions" WHERE "versions"."full_name" IS NULL LIMIT $1  [["LIMIT", 1]]
  TRANSACTION (0.4ms)  ROLLBACK
Traceback (most recent call last):
        3: from (irb):3
        2: from (irb):4:in `rescue in irb_binding'
        1: from lib/certificate_chain_serializer.rb:13:in `dump'
NoMethodError (undefined method `map' for #<String:0x00005607a650ce78>)
Did you mean?  ta
```

Reply:
nevermind, we can ignore these. it is covered by Gem::Package.new. the following works fine.
```
  test "push a new version of with invalid cert" do
    rubygem = create(:rubygem, name: "sandworm", number: "1.0.0")
    create(:ownership, rubygem: rubygem, user: @user)

    build_gem "sandworm", "2.0.0" do |gemspec|
      gemspec.cert_chain = ["notcert"]
    end


    push_gem "sandworm-2.0.0.gem"
    assert_response :unprocessable_entity
  end
  ```